common/core/gpudevice.cpp

```diff
@@ -307,6 +307,435 @@ void GpuDevice::PanoramaInit(
+        break;
+      }
+    }
+    if (!skip_duplicate_display) {
+      // save the sub display num for the mosaic display (don't care
+      // if
+      // the physical/logical display is existing/connected here)
+      clone_display.emplace_back(i_clone_split_num);
+      clone_duplicate_check.emplace_back(i_clone_split_num);
+    }
+  }
+  cloned_displays.emplace_back(clone_display);
+}
+
+void GpuDevice::ParsePhysicalDisplayRotation(
+    std::string &value, std::vector<uint32_t> &display_rotation,
+    std::vector<uint32_t> &rotation_display_index) {
+  std::istringstream i_value(value);
+  std::string physical_index_str;
+  std::getline(i_value, physical_index_str, ':');
+  if (physical_index_str.empty() ||
+      physical_index_str.find_first_not_of("0123456789") != std::string::npos)
+    return;
+
+  uint32_t physical_index = atoi(physical_index_str.c_str());
+  // Check and skip if the display is already in use.
+  bool skip_duplicate_display = false;
+  size_t rotation_size = rotation_display_index.size();
+  for (size_t i = 0; i < rotation_size; i++) {
+    if (rotation_display_index.at(i) == physical_index) {
+      skip_duplicate_display = true;
+      break;
+    }
+  }
+
+  if (skip_duplicate_display) {
+    return;
+  }
+
+  std::string rotation_str;
+  // Got split num
+  std::getline(i_value, rotation_str, ':');
+  if (rotation_str.empty() ||
+      rotation_str.find_first_not_of("0123") != std::string::npos)
+    return;
+
+  uint32_t rotation_num = atoi(rotation_str.c_str());
+  display_rotation.emplace_back(rotation_num);
+  rotation_display_index.emplace_back(physical_index);
+}
+
+void GpuDevice::ParseFloatDisplaySetting(
+    std::string &value, std::vector<HwcRect<int32_t>> &float_displays,
+    std::vector<uint32_t> &float_display_indices) {
+  std::istringstream i_value(value);
+  std::string index_str;
+  std::string float_rect_str;
+  std::vector<int32_t> float_rect;
+
+  // Got display index
+  std::getline(i_value, index_str, ':');
+  if (index_str.empty() ||
+      index_str.find_first_not_of("0123456789") != std::string::npos) {
+    return;
+  }
+
+  int32_t index = atoi(index_str.c_str());
+
+  // Got rectangle configuration
+  while (std::getline(i_value, float_rect_str, '+')) {
+    if (float_rect_str.empty() ||
+        float_rect_str.find_first_not_of("0123456789") != std::string::npos) {
+      continue;
+    }
+    size_t float_rect_val = atoi(float_rect_str.c_str());
+
+    // Save the rectangle - left, top, right & bottom
+    float_rect.emplace_back(float_rect_val);
+  }
+
+  // If entire rect is available, store the parameters
+  // TODO remove hard code
+  if (float_rect.size() == 4) {
+    float_display_indices.emplace_back(index);
+    HwcRect<int32_t> rect = HwcRect<int32_t>(
+        float_rect.at(0), float_rect.at(1), float_rect.at(2), float_rect.at(3));
+    float_displays.emplace_back(rect);
+  }
+}
+
+void GpuDevice::InitializeDisplay(std::vector<uint32_t> &physical_displays,
```

Comment:
Can we change the name?
IntializeDisplayOrder?

Reply:
Or IntiailizeDisplayIndex

Reply:
Sure, will change accordingly.